substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/graal/code/SubstrateBackend.java

```diff
@@ -152,14 +166,220 @@ public CompilationResult newCompilationResult(CompilationIdentifier compilationI
         return new CompilationResult(compilationIdentifier, name) {
             @Override
             public void close(OptionValues options) {
-                /*
-                 * Do nothing, we do not want our CompilationResult to be closed because we
-                 * aggregate all data items and machine code in the native image heap.
-                 */
+                FrameOptimizer.compressInfoPoints(this);
```

Comment:
A better place to to do the deduplication of BytecodeFrames (incrementally) would be to add it to `org.graalvm.compiler.code.CompilationResult#addInfopoint` and the again later when there is no more mutation in the `jdk.vm.ci.code.BytecodeFrame#values` array.

Its very unfortunate that we have a mutable state in `BytecodeFrame` while at the same time the `jdk.vm.ci.code.DebugInfo#bytecodePosition` field is final. This forces us to instantiate new Infopoints to be able to have that `bytecodePosition` field updated. This creates new problems since the infopoint instances are at that point already referenced at other places as well requiring shenanigans like in line 250. @adinn @christianwimmer

Reply:
@olpaw @christianwimmer Yes, my first attempt was to deduplicate `BytecodeFrame` instances when they are created so that we never incur the penalty of modelling what we know will actually be a tree as a set of paths. I quickly found that they are not stable at create time so had to pick a later point to convert from one representation to the other.

I chose to do it in (the Substrate override of) `CompilationResult.close` because that means the resulting footprint compression is pipelined with compilation rather than delayed until compilation completes. However, that choice requires wholesale replacement of the set of Infopoint instances in `CompilationResult.infopoints` after the collection has been assembled, at which point references to the old values may have been stashed away -- e.g in list `CompilationResult.invalidCallDeoptimizationStates`. Doing it earlier, say at the call to `addInfopoint`,  could mean we can swap in a replacememt before it is used elsewhere.

@olpaw I am not sure what you mean by this:

> A better place to do the deduplication of BytecodeFrames (incrementally) would be to add it to org.graalvm.compiler.code.CompilationResult#addInfopoint and then again later when there is no more mutation in the jdk.vm.ci.code.BytecodeFrame#values array.

Does the `values` array get modified after the call to `addInfopoint`? If so then how can we deduplicate frames? at the call? Will we not risk eliding two frames that later turn out to be different because they end up with different values?

n.b. I have not been able to find any use of an `Infopoint` added via a call to `addInfopoint` other than via `CompilationResult.recordCall`. However, in all the latter cases the `Call` returned by this seems only to be used to update `CompilationResult.invalidCallDeoptimizationStates`. Is there some other use of a newly added `Infopoint` that I am missing?

Also, I have run command ` mx --env ce-js build` about a dozen times and not been able to see an error. Is the problem very infrequent or could the failure to manifest be in some way related to my setup?